Vault aws iam auth method

semiria opened the pull request:
Fixes #36 

Allow AWS IAM auth method to authenticate to Vault

sguilbert-sfdc committed:
Add AWS IAM auth method to authenticate to Vault

sguilbert-sfdc committed:
Add documentation to cover AWS IAM Auth method

VladRassokhin commented on server/src/main/resources/buildServerResources/editProjectConnectionVault.jsp:
```diff
@@ -163,10 +163,19 @@
+
   $j('#aws-iam-auth').click(function() {
       $j("#approle_auth_endpoint").toggle(!this.checked);
+      $j("#approle_auth_endpoint input").val("")
```
Please don't, it empties unsaved user input if someone accidentally clicked on that checkbox.
That should be done on server side

VladRassokhin commented on build.gradle:
```diff
@@ -56,8 +56,8 @@ subprojects {
 
-    sourceCompatibility = "1.6"
-    targetCompatibility = "1.6"
+    sourceCompatibility = "1.8"
```
Is that really necessary?

semiria replied:
Yes for it to work on my development environment, feel free to put that back to 1.6.

VladRassokhin commented on server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultConnector.kt:
```diff
@@ -105,7 +109,35 @@ class VaultConnector(dispatcher: EventDispatcher<BuildServerListener>, private v
+            val options = AwsIamAuthenticationOptions.builder()
+                    .credentialsProvider(InstanceProfileCredentialsProvider.getInstance()).build()
+
+            val template = createRestTemplate(settings, trustStoreProvider)
```
Probably original `template` from `getToken(...)` could be used, no need to create another one

VladRassokhin commented on server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultConnector.kt:
```diff
@@ -105,7 +109,35 @@ class VaultConnector(dispatcher: EventDispatcher<BuildServerListener>, private v
+
+                return token.token to accessor
+            } catch (e: Exception) {
+                throw Exception(e)
```
Hm?

VladRassokhin commented on server/src/main/resources/buildServerResources/editProjectConnectionVault.jsp:
```diff
@@ -153,4 +162,20 @@
+  })
+
+  $j('#aws-iam-auth').click(function() {
+      $j("#approle_auth_endpoint").toggle(!this.checked);
```
Could be simplified by using class `auth_approle` on all relevant `tr` tags and `$j(".auth_approle").toggle(!this.checked)`

VladRassokhin commented on server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultProjectConnectionProvider.kt:
```diff
@@ -64,14 +64,18 @@ class VaultProjectConnectionProvider(private val descriptor: PluginDescriptor) :
-                if (it[VaultConstants.FeatureSettings.SECRET_ID].isNullOrBlank()) {
-                    errors.add(InvalidProperty(VaultConstants.FeatureSettings.SECRET_ID, "Should not be empty"))
+
+                val awsIamAuth = it[VaultConstants.FeatureSettings.AWS_IAM_AUTH] ?: "false"
```
Please change to `?.toBoolean() ?: false` as it's used elsewhere (grep `toBoolean` on project)

VladRassokhin reviewed (changes requested):
Thanks for your changes, please take a look on comments and edit commits history using interactive rebase.

Also I'm wondering whether it would be better to have radio buttons: `AppRole Auth` and `AWS IAM Auth`

Although I could finish it if it's OK and you don't want to invest more time

semiria commented:
Thank you for your feedback @VladRassokhin. I've addressed most of your comments and left feedback on what I didn't. I think the radio button versus checkbox is a good idea, I went ahead and modified the code to reflect that comment.

I've sent you an invite to contribute on this fork, we'd be ok if you want to take over from here.

semiria commented:
Any feedback @VladRassokhin ? As mentioned you can take over from here or I can address any other comments you may have, I'm open to both options.
Thanks!

VladRassokhin commented:
Thank you for contribution, I've rebased, refactored and manually merged into master. Please try [latest build](https://teamcity.jetbrains.com/guestAuth/repository/download/TeamCityPluginsByJetBrains_TeamCityHashiCorpVaultPlugin_Build/.lastSuccessful/teamcity-hashicorp-vault-plugin.zip).

semiria commented:
Thank you so much @VladRassokhin , I'll let you know if something comes up
